// @rcade/sdk/dist/index.js
// src/plugin_channel/index.ts
class PluginChannel {
  port;
  channel;
  pendingRequests = new Map;
  static async acquire(name, version) {
    const nonce = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
    return new Promise((resolve, reject) => {
      const listener = (event) => {
        if (event.data?.type === "plugin_channel" && event.data?.nonce === nonce) {
          const port = event.ports[0];
          if ("error" in event.data) {
            window.removeEventListener("message", listener);
            reject(new Error(event.data.error));
            return;
          }
          const { channel } = event.data;
          if (!channel?.name || !channel?.version || !port) {
            console.warn("Invalid plugin_channel event", event.data);
            return;
          }
          window.removeEventListener("message", listener);
          const pluginChannel = new PluginChannel(port, channel);
          resolve(pluginChannel);
        }
      };
      window.addEventListener("message", listener);
      window.parent.postMessage({
        type: "acquire_plugin_channel",
        nonce,
        channel: { name, version }
      }, "*");
    });
  }
  constructor(port, channel) {
    this.port = port;
    this.channel = channel;
    this.port.addEventListener("message", (event) => {
      const { _nonce } = event.data ?? {};
      if (_nonce && this.pendingRequests.has(_nonce)) {
        const pending = this.pendingRequests.get(_nonce);
        this.pendingRequests.delete(_nonce);
        pending.resolve(event.data);
      }
    });
    this.port.start();
  }
  getPort() {
    return this.port;
  }
  getVersion() {
    return String(this.channel.version);
  }
  request(message, timeoutMs = 5000) {
    const _nonce = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pendingRequests.delete(_nonce);
        reject(new Error("Request timed out"));
      }, timeoutMs);
      this.pendingRequests.set(_nonce, {
        resolve: (data) => {
          clearTimeout(timeout);
          resolve(data);
        },
        reject: (error) => {
          clearTimeout(timeout);
          reject(error);
        }
      });
      this.port.postMessage({ ...message, _nonce });
    });
  }
}
export {
  PluginChannel
};
